### PR TITLE
Implement popupmenu as a floating grid internally to reduce flicker

### LINKED ISSUE #9542: E41: Out of memory when `rightleft` is set and a certain plugin is used
- `nvim --version`:
```
NVIM v0.3.4
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.3.4/src -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```
- Vim (version: ) behaves differently? Yes but it's Irrelevant because the said plugin operates differently in Vim and Neovim,
- Operating system/version:
```
$ uname -a
Linux NUC 4.20.3-arch1-1-ARCH #1 SMP PREEMPT Wed Jan 16 22:38:58 UTC 2019 x86_64 GNU/Linux
```
- Terminal name/version: rxvt-unicode
- `$TERM`: `screen` Inside Tmux

### Steps to reproduce

```
mkdir -p ~/.config/nvim/pack/crash-test/start
git clone https://github.com/neoclide/coc.nvim ~/.config/nvim/pack/crash-test/start/coc
cd ~/.config/nvim/pack/crash-test/start/coc
git checkout 9aaeaa40a50819008c217cd3e31857b96c475f4d
yarn install
nvim
```

Insert the following text to the buffer:

```
שלום bla,

On Sat, Jan 12, 2019 at 10:21:38AM +0200, Someone wrote:

> היי דורון,
> להוסיף עוד 115 ש״ח הוא נכון.
> 
> תודה וסליחה על אי ההבנה,
> bla
> 
>
```

```
:set rightleft
" open a new line at the end and try to enter text, for example, use the following keys:
Gos
```

### Actual behaviour

Crash.

### Expected behaviour

No Crash.

### Additional information

A bug report was added to the plugin's project here - https://github.com/neoclide/coc.nvim/issues/369 and it was closed but not really fixed but sort of avoided. We both have no idea how this bug is produced, even with `nvim -Vcoc-crash.debug` no interesting output is available in the log.

### PR DESCRIPTION
**Pull request by bfredl**
Extract the "compositor" layer from #6619. This allows us to draw the popupmenu on a separate grid rather than on the default grid, even for TUI use.

Attempt to eliminate flicker in all these cases:

- The pum was filtered due to the user added a prefix char. 
- The pum is moved due to linewrap change ( i e by long inserted item at end of line)
- The contents of the pum was changed due to `complete()` call or something (corresponding to https://github.com/vim/vim/commit/ae654385dfb2ae4c1d70789d1dce3676dba4dfbc, tested now)

Also `update_screen(NOT_VALID)` should never be needed to redraw the text under the pum, rather the cached contents is resent from `default_grid`.

Note: in principle this will enable two connected UI:s, one using `ext_popupmenu` and the other using the builtin popupmenu, at the same time. This will need some other refactors though.

Note 2: this should reduce the bookkeeping in `edit.c` for pum redraws (ideally all commented code can be removed, but needs more testing). This will lower the bar for `ex_getln.c` also using the pum instead of statusline wildmenu for completions. It will need to compute position differently, but apart from that it would be able to use `pum_display` and `pum_undisplay` without thinking about window redraws at all.

fixes #9542.

### CODE REVIEW

**Comment by bfredl** [comment]
ping @Shougo what is the typical situation for https://github.com/vim/vim/commit/ae654385dfb2ae4c1d70789d1dce3676dba4dfbc, is it multiple `complete()` calls changing the popupmenu?

File: src/nvim/compositor.c [hunk 1]
```diff
@@ -0,0 +1,369 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
```
Comment (justinmk) [inline src/nvim/compositor.c]: name it `ui_compositor.c` , so it is grouped with `ui.c` and `ui_bridge.c` ? And the prefix could be `uicomp_`. This helps discover the project structure.

File: src/nvim/ui.c [hunk 2]
```diff
@@ -338,6 +359,18 @@ void ui_line(ScreenGrid *grid, int row, int startcol, int endcol, int clearcol,
   }
 }
 
+void ui_composed_call_raw_line(Integer draw_grid, Integer row,
```
Comment (justinmk) [inline src/nvim/ui.c]: isn't "composite" the more common phrasing rather than "composed"?

File: src/nvim/ui.c [hunk 3]
```diff
@@ -85,21 +86,25 @@ static char uilog_last_event[1024] = { 0 };
 //
 // See http://stackoverflow.com/a/11172679 for how it works.
 #ifdef _MSC_VER
-# define UI_CALL(funname, ...) \
+# define UI_CALL_CND(CND, funname, ...) \
```
Comment (justinmk) [inline src/nvim/ui.c]: what does CND mean?

**Comment by Shougo** [comment]
> is it multiple complete() calls changing the popupmenu?

Yes.  I will test the branch later.

File: src/nvim/ui.c (hunk 3 above)
Comment (bfredl) [reply src/nvim/ui.c]: condition, maybe `_IF` is better?

File: src/nvim/ui.c (hunk 2 above)
Comment (bfredl) [reply src/nvim/ui.c]: Maybe, but "composed" is a passive/past tense as in "the UI is (being) composed" or "after the line was composed"

File: src/nvim/compositor.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/compositor.c]: yeah, lets do that.

**Comment by bfredl** [comment]
And another thing: using compositior for pum also make possible things like [this](https://github.com/bfredl/neovim/commit/9849e66eecc715ba1981b3f44b0595cb0edbe30c):
screenshot from 2019-01-20 21-27-23

Probably not practially useful for anything (except I guess providing useful context once in a blue moon), but maybe it can help make up for the lack of `:smile`.

File: src/nvim/ui.c (hunk 3 above)
Comment (justinmk) [reply src/nvim/ui.c]: FWIW we use `cond` in a few places.

**Comment by bfredl** [comment]
Thanks, I will investigate.

**Comment by bfredl** [comment]
@shougo I think at least some of the flicker here is fundamental, not related to this patch. Because if you press a key that changes the pum, and then 0-30ms later a timer will change the pum back again, there _will_ be flicker for up to 30ms. Adding this line fixes it:

```
   autocmd TextChangedP * call s:check_changes()
```

Does nvim behave subtly different than vim for the behavior of autocmds, perhaps?

**Comment by bfredl** [comment]
But I changed timer to manually invoked `<cmd>` command and there is still some flicker, so all was not timer related.

**Comment by bfredl** [comment]
Hmm state management in `edit.c` is a bit more complicated than I thought. At several points where the pum is undisplayed, the code doesn't know if the pum is going to be displayed immediately again. (The code might say `pum_undisplay(); ins_show_completion_again()` but the later sometimes ends up not showing the pum again). I will try a different approach:

- at every point `pum_undisplay` or `pum_display` is called, just store the internal state
- do the actual pum draw or clearing in `ins_redraw`.

**Comment by bfredl** [comment]
Did the defer for only undisplay for now, probably enough.

@shougo please try it again.

**Comment by Shougo** [comment]
I have tested the latest version.
I have confirmed the flicker problem is fixed.
Thank you.

**Comment by bfredl** [comment]
Issue: invoke the popupmenu with lots of items so it covers the entire the height of the screen. Now do `:<c-r>=<c-f>`. The cmdline window is now unusable due to covered by the pum. But the same issue exits on vim master, I will report it upstream.

**Comment by bfredl** [comment]
vim issue: https://github.com/vim/vim/issues/3838

**Comment by bfredl** [comment]
Update: treatment of doublewidth chars is a huge mess, but is hopefully working now.

**Comment by bfredl** [comment]
I added a bunch of tests for popupmenu redrawing in various situation, including making pum bigger/smaller, and move in all four directions, as well as interaction with non-current window scroll and multiline messages.

I will dogfeed this branch as my default nvim for some time, to check for further assertion errors.

**Comment by Shougo** [comment]
I have tested the branch, and it is faster completion than master.

**Comment by bfredl** [comment]
Made some simplifications to the dispatching logic. A single `UI_CALL` macro now does all the work without a bunch of helper macros for `__VA_ARGS__` shenanigans. Also internal-only events use `ui_call_evname` wrappers, just like internal+external and external-only events.

File: src/nvim/edit.c [hunk 4]
```diff
@@ -2490,28 +2493,18 @@ void set_completion(colnr_T startcol, list_T *list)
 static pumitem_T *compl_match_array = NULL;
 static int compl_match_arraysize;
 
+
 /*
  * Update the screen and when there is any scrolling remove the popup menu.
  */
```
Comment (justinmk) [inline src/nvim/edit.c]: orphan comment

**Comment by purpleP** [comment]
Seems like you have a bug. After filtering spaces between items and metadata disappear.

**Comment by bfredl** [comment]
@purpleP yes, the metadata seems to be shifted to the left one cell. Thanks for noticing, I will fix it.

**Pull request merged**